### PR TITLE
Fix Issue #1039 : extending KeyboardShortcut to Support Special Keys

### LINKED ISSUE #1039: Chat / The escape key should close open modals
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___

**Platform - version:**  
Desktop, build 280

**Action Performed (reproducible steps):**

Test 1
1. Click the paperclip icon
2. Select a file to upload
3. Before confirming the upload, press Escape
4. Observe that the upload modal stays on the screen

Test 2
1. Click the paperclip icon
2. Select an image to upload
3. Upload the file
4. Click into the image 
5. Press escape
6. Observe that the attachment modal stays on the screen

**Expected Result:**
All modals should be cancelled when the user presses the Escape key. 

**Actual Result:**
The modal stays on the screen

**Notes/Photos/Videos:** 
102928813-6878ce80-4467-11eb-8aec-82c137cbf461

view the job on Upwork [here](https://www.upwork.com/jobs/~01122a92e0900bb65c).

### PR DESCRIPTION
**Pull request by f0wu5u**
<If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit.>

### Details
Added Escape key listener to close all modals using existing KeyboardShortcut library. KeyboardShortcut library was also modified to accept single keys by passing `special` as modifier when subscribing for event.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify.cash/issues/1039

### Tests

Close Image upload preview:

1. Go to chat screen of Expensify.cash
2. Click on image upload button and select image for preview
3. Press Escape button and modal should close

Close Attachment preview modal:
1. Press on any image in chat
2. Wait for preview modal
3. Press Escape and modal should close

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### Screenshots

https://user-images.githubusercontent.com/26269088/103719406-e0dfa380-4fc0-11eb-9bcc-f145b7ed4377.mp4

### CODE REVIEW

**Comment by code-fi** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by alex-mechler** [comment]
Thanks for the PR @code-fi! Can you resolve the merge conflicts?

**Comment by code-fi** [comment]
@alex-mechler merge conflicts resolved :)

**Comment by Jag96** [comment]
@code-fi did you open another PR for this? Just double-checking since I don't see a new PR linked here or in the issue

**Comment by code-fi** [comment]
@Jag96 no, it was closed automatically I think

File: src/components/Modal.js [hunk 1]
```diff
@@ -39,6 +40,12 @@ const Modal = (props) => {
         needsSafeAreaPadding,
     } = getModalStyles(props.type, useWindowDimensions());
 
+    //register escape key listener via effect
```
Comment (alex-mechler) [inline src/components/Modal.js:43]: It looks like the last style run timed out, but this should have a space between `//` and the comment, and start with a capital letter.

File: src/libs/KeyboardShortcut/index.js [hunk 2]
```diff
@@ -7,15 +7,16 @@ const events = {};
...
         return;
     }
 
     // The active callback is the last element in the array
-    const eventCallbacks = events[event.keyCode];
+    const eventCallbacks = events[event.key];
     const callback = eventCallbacks[eventCallbacks.length - 1];
 
     const pressedModifiers = _.all(callback.modifiers, (modifier) => {
+
```
Comment (alex-mechler) [inline src/libs/KeyboardShortcut/index.js:19]: Why was this line added?

File: src/libs/KeyboardShortcut/index.js [hunk 3]
```diff
@@ -74,20 +73,23 @@ const KeyboardShortcut = {
      * @param {Boolean} captureOnInputs Should we capture the event on inputs too?
      */
     subscribe(key, callback, modifiers = 'shift', captureOnInputs = false) {
-        const keyCode = key.charCodeAt(0);
-        if (events[keyCode] === undefined) {
-            events[keyCode] = [];
+
+        //enable support for special keys like Escape
+        //const keyCode =  key.charCodeAt(0);
```
Comment (alex-mechler) [inline src/libs/KeyboardShortcut/index.js:78]: Remove commented out code like this.

**Review by alex-mechler — CHANGES_REQUESTED** [review]
I think there are some style issues, but github was having issues when you last pushed. You can run the style checker locally with `npm run lint`

**Comment by code-fi** [comment]
@alex-mechler changes made

File: src/libs/KeyboardShortcut/index.js [hunk 4]
```diff
@@ -74,19 +72,20 @@ const KeyboardShortcut = {
      * @param {Boolean} captureOnInputs Should we capture the event on inputs too?
      */
     subscribe(key, callback, modifiers = 'shift', captureOnInputs = false) {
-        const keyCode = key.charCodeAt(0);
-        if (events[keyCode] === undefined) {
-            events[keyCode] = [];
+        // Enable support for special keys like Escape
+
```
Comment (alex-mechler) [inline src/libs/KeyboardShortcut/index.js:76]: Extra line here can be removed.

**Review by alex-mechler — CHANGES_REQUESTED** [review]
When pressing `escape`, we get an outline around an element that should not be there. 
image

File: src/libs/KeyboardShortcut/index.js (hunk 4 above)
Comment (code-fi) [reply src/libs/KeyboardShortcut/index.js]: Unfortunately I couldn't reproduce issue with outline at my end... can you state platform and OS you used?

**Comment by alex-mechler** [comment]
@code-fi pulling this out of the thread, I was able to reproduce on Mac OS (10.15.7), on both desktop and web

**Comment by Jag96** [comment]
Closing in favor of https://github.com/Expensify/Expensify.cash/pull/1293
